linux/hid.c

```diff
@@ -130,6 +135,48 @@ static wchar_t *utf8_to_wchar_t(const char *utf8)
 	return ret;
 }
 
+
+/* Set the last global error to be reported by hid_error(NULL).
+ * The given error message will be copied (and decoded according to the
+ * currently locale, so do not pass in string constants).
+ * The last stored global error message is freed.
+ * Use register_global_error(NULL) to indicate "no error". */
+static void register_global_error(const char *msg)
+{
+	if (last_global_error_str)
+		free(last_global_error_str);
+
+	last_global_error_str = msg? utf8_to_wchar_t(msg): NULL;
+}
+
+
+/* Set the last error for a device to be reported by hid_error(device).
+ * The given error message will be copied (and decoded according to the
+ * currently locale, so do not pass in string constants).
+ * The last stored global error message is freed.
+ * Use register_device_error(device, NULL) to indicate "no error". */
+static void register_device_error(hid_device *dev, const char *msg)
+{
+	if (dev->last_error_str)
+		free(dev->last_error_str);
+
+	dev->last_error_str = msg? utf8_to_wchar_t(msg): NULL;
```

Comment:
`msg` will be evaluated to `false` when you call `register_device_error(NULL);` so I think you could change
```
dev->last_error_str = msg? utf8_to_wchar_t(msg): NULL;
```
to 
```
dev->last_error_str = msg;
```

Reply:
Sorry, I forgot about need to call utf8_to_wchar_t. So the `? :` is needed.

Reply:
Applied
```diff
diff --git a/linux/hid.c b/linux/hid.c
index 6ce2d9e..4b6cad3 100644
--- a/linux/hid.c
+++ b/linux/hid.c
@@ -146,7 +146,7 @@ static void register_global_error(const char *msg)
        if (last_global_error_str)
                free(last_global_error_str);
 
-       last_global_error_str = msg? utf8_to_wchar_t(msg): NULL;
+       last_global_error_str = utf8_to_wchar_t(msg);
 }
 
 
@@ -160,7 +160,7 @@ static void register_device_error(hid_device *dev, const char *msg)
        if (dev->last_error_str)
                free(dev->last_error_str);
 
-       dev->last_error_str = msg? utf8_to_wchar_t(msg): NULL;
+       dev->last_error_str = utf8_to_wchar_t(msg);
 }
 
 /* See register_device_error, but you can pass a format string into this function. */
```

Reply:
looks good